src/main/java/com/aws/greengrass/componentmanager/plugins/DockerImageDownloader.java

```diff
@@ -188,4 +137,107 @@ public boolean canUnarchiveArtifact() {
         // Not applicable for docker images since docker engine abstracts this
         return false;
     }
+
+    private void checkDownloadPrerequisites() throws PackageDownloadException {
+        // Check that Docker engine is installed
+        if (!dockerClient.dockerInstalled()) {
+            throw new PackageDownloadException("Docker engine is not installed on the device, please ensure it's "
+                    + "installed and redo the deployment");
+        }
+    }
+
+    private File performDownloadSteps(Image image) throws PackageDownloadException, InterruptedException {
+        AtomicBoolean credentialRefreshNeeded = new AtomicBoolean(false);
+        do {
+            // Attempt getting credentials and login only for public registries, currently we only
+            // support private registries in ECR so assume others are public.
+            if (image.getRegistry().isEcrRegistry() && image.getRegistry().isPrivateRegistry()) {
+                // Get auth token for ECR
+                try {
+                    RetryUtils.runWithRetry(infiniteAttemptsRetryConfig, () -> {
+                        Registry.Credentials credentials =
+                                ecrAccessor.getCredentials(image.getRegistry().getRegistryId());
+                        image.getRegistry().setCredentials(credentials);
+                        credentialRefreshNeeded.set(false);
```

Comment:
Since each image has a separate instance of the registry, if you have multiple artifacts from the same ECR registry will this get credentials multiple times?

Reply:
Yes, and this is not ideal, but that would need to be maintained across all artifacts within a deployment or even across deployments. Right now the artifact downloaders get instantiated for each unique artifact and don't share context. So that will need significant refactor, I have a TODO in the backlog for optimizing that and maintaining some sort of cache